clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -0,0 +1,1064 @@
+         * even be in the process of being processed by the remote server.  At the moment, our only option
+         * to time them out is to close the entire connection.
+         *
+         * @param now                   The current time in milliseconds.
+         * @param callsInFlight         A map of nodes to the calls they have in flight.
+         */
+        private void timeoutCallsInFlight(long now, Map<String, List<Call>> callsInFlight) {
+            int numTimedOut = 0;
+            for (Map.Entry<String, List<Call>> entry : callsInFlight.entrySet()) {
+                List<Call> contexts = entry.getValue();
+                if (contexts.isEmpty())
+                    continue;
+                String nodeId = entry.getKey();
+                // We assume that the first element in the list is the earliest.  So it should be the
+                // only one we need to check the timeout for.
+                Call call = contexts.get(0);
+                if (calcTimeoutMsRemainingAsInt(now, call.deadlineMs) < 0) {
+                    log.debug("{}: Closing connection to {} to time out {}", clientId, nodeId, call);
+                    client.close(nodeId);
+                    numTimedOut++;
+                    // We don't remove anything from the callsInFlight data structure.  Because the connection
+                    // has been closed, the calls should be returned by the next client#poll(),
+                    // and handled at that point.
+                }
+            }
+            if (numTimedOut > 0)
+                log.debug("{}: timed out {} call(s) in flight.", clientId, numTimedOut);
+        }
+
+        /**
+         * Handle responses from the server.
+         *
+         * @param now                   The current time in milliseconds.
+         * @param responses             The latest responses from KafkaClient.
+         * @param correlationIdToCall   A map of correlation IDs to calls.
+         * @param callsInFlight         A map of nodes to the calls they have in flight.
+        **/
+        private void handleResponses(long now, List<ClientResponse> responses, Map<String, List<Call>> callsInFlight,
+                Map<Integer, Call> correlationIdToCall) {
+            for (ClientResponse response : responses) {
+                int correlationId = response.requestHeader().correlationId();
+
+                Call call = correlationIdToCall.get(correlationId);
+                if (call == null) {
+                    // If the server returns information about a correlation ID we didn't use yet,
+                    // an internal server error has occurred.  Close the connection and log an error message.
+                    log.error("Internal server error on {}: server returned information about unknown " +
+                        "correlation ID {}", response.destination(), correlationId);
+                    client.close(response.destination());
```

Comment:
We don't know what request the response corresponds to, because its correlation ID is invalid.  Of course, requests that don't receive a response will time out after enough time.